### PR TITLE
feat(lib): make calls with ignored access easier

### LINKED ISSUE #6694: Rework access overrides api
While talking with @juho-jaakkola about fixing missing and badly applied ignore access and show hidden entities flags I got an idea to adjust it's api. Current problems:
- ignore access and hidden entities have different api that makes use of it one line longer than necessary and unnecessairly convoluted
- it's easy to forget about hidden entities in relevant cases
- it's easy to forget to revoke raised access
- it's easy to switch access back to wrong value (typo in var name) or just lower access by using [hardcoded value instead of var](https://github.com/Elgg/Elgg/blob/master/engine/lib/upgrades/2010052601.php#L27)

Those bugs are likely to cause critical security issue, so if we can improve it somehow It's most likely worth it.

One if the ideas I have would be to introduce access stack (as service) that would work similar to:

``` php
elgg_access()->pushIgnoreAccess(true);

//do stuff

elgg_access()->popIgnoreAccess();
```

or

``` php
// elgg_access()->pushSession() would return some kind of "access session" object - TBD
elgg_access()->pushSession()->setIgnoreAccess(true)->setShowHidden(true);

//do stuff

elgg_access()->popSession();
```

Optional wishlist:
- allow injecing session on top of existing one, ie. I'm an admin, but want to have a tool that injects test user session on top of mine for access testing, but allows to pop current one and go back to undelying one. That's a stretch and not entirely consistent with current feature that works in local context, but might be something to approach in future.
- allow easily mocking sessions for tests of entity access

Thoughts?

### PR DESCRIPTION
**Pull request by hypeJunction**
Adds a utility function to call methods and functions with
ignored access and showing hidden entities

Fixes #6694

### CODE REVIEW

File: engine/lib/elgglib.php [hunk 1]
```diff
@@ -1211,6 +1211,39 @@ function elgg_extract_class(array $array, $existing = [], $extract_key = 'class'
...
+/**
+ * Calls a callable passing arguments and applying logic based on flags
+ *
+ * @param callable $callable  Callable to call
+ * @param array    $arguments Arguments to pass to callable
+ * @param null     $flags     Flags
+ *
+ * @return mixed
+ */
+function elgg_call(callable $callable, $arguments = null, $flags = null) {
```
Comment (mrclay) [inline engine/lib/elgglib.php]: I'd say make $flags required and 1st argument, and no $argument. We really don't want devs to replace static analysis-friendly function calls with callbacks.

File: engine/lib/elgglib.php [hunk 2]
```diff
@@ -1211,6 +1211,39 @@ function elgg_extract_class(array $array, $existing = [], $extract_key = 'class'
...
+	}
+
+	$ha = access_get_show_hidden_status();
+	if ($flags & ELGG_SHOW_DISABLED_ENTITIES) {
+		access_show_hidden_entities(true);
+	} else if ($flags & ELGG_HIDE_DISABLED_ENTITIES) {
+		access_show_hidden_entities(false);
+	}
+
+	$result = call_user_func_array($callable, (array) $arguments);
```
Comment (mrclay) [inline engine/lib/elgglib.php]: For this to be bulletproof you need to catch exceptions, reset access, and I guess re-throw them.

File: engine/lib/elgglib.php [hunk 3]
```diff
@@ -1211,6 +1211,50 @@ function elgg_extract_class(array $array, $existing = [], $extract_key = 'class'
 	return array_values(array_unique($existing));
 }
 
+/**
+ * Calls a callable passing arguments and applying logic based on flags
+ *
+ * @param null    $flags   Flags
```
Comment (mrclay) [inline engine/lib/elgglib.php]: int.

Comment (mrclay) [reply engine/lib/elgglib.php]: We can use the int type-hint, too.

File: engine/classes/ElggEntity.php [hunk 4]
```diff
@@ -1622,38 +1620,33 @@ public function disable($reason = "", $recursive = true) {
...
-				$options[$db_column] = $guid;
-
-				$subentities = new \ElggBatch('elgg_get_entities', $options);
-				$subentities->setIncrementOffset(false);
-
-				foreach ($subentities as $subentity) {
-					/* @var $subentity \ElggEntity */
-					if (!$subentity->isEnabled()) {
-						continue;
+			$disable_relatives = function () use ($guid, $reason) {
```
Comment (mrclay) [inline engine/classes/ElggEntity.php]: I think when we define the callable like this, we should define $flags directly above it, to help readers understand the environment in which it'll run.

**Review by jeabakker — APPROVED** [review]

**Comment by jdalsem** [comment]
i am ambiguous about this feature... does it need docs in developer guides somewhere?

rebase to merge

**Comment by hypeJunction** [comment]
Ambigous or ambivolent?
Let's update the core to use this first, then we can advise plugins to use it.

**Comment by jdalsem** [comment]
both... but ambivalent the most i guess :)

**Comment by jdalsem** [comment]
it probably needs to grow on me :)

**Pull request merged**